File: app/src/main/java/com/example/thinkdo/fragment/ConstraintFragment.java
package com.example.thinkdo.fragment;

import android.os.Bundle;
import androidx.annotation.Nullable;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.example.thinkdo.compoentdemo.R;

import butterknife.BindView;

public class ConstraintFragment extends BaseFragment implements View.OnClickListener {
    EditText et;
    Button btn;

    @BindView(R.id.textView4)
    TextView textView4;

    @Override
    public int getLayoutId() {
        return R.layout.constraint_test;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        et = view.findViewById(R.id.editText);
        btn = view.findViewById(R.id.button);

//        et.setFocusable(false);
//        et.setFocusableInTouchMode(false);

        et.setOnClickListener(this);
        btn.setOnClickListener(this);

        test();
    }



    private void test(){

//
//        StringBuilder builder = new StringBuilder();
//        for (int i = 127; i < 200; i++) {
//            byte b = (byte)i;
//
//            byte[] bytes = new byte[2];
//            bytes[0] = b;
//            bytes[1] = 0x2c;
//            String str = new String(bytes, new Char);
//            builder.append(i+":"+str+"\n");
//        }
//
//        textView4.setText(builder.toString());
    }


    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.editText:
                Log.d("TAG","editText");
                break;
            case R.id.button:
                Log.d("TAG","button");
                break;
        }
    }
}

File: kotlin/src/main/java/com/example/xh/kotlin/fragment/ConstraintFragment.kt
package com.example.xh.kotlin.fragment

import android.graphics.Rect
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.xh.kotlin.BaseFragement
import com.example.xh.kotlin.R
import kotlinx.android.synthetic.main.fragment_constraint_layout.*
import kotlinx.android.synthetic.main.recycler_constraint_item.*

class ConstraintFragment : BaseFragement() {

    override fun getLayoutResId(): Int {
        return R.layout.fragment_constraint_layout
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)


        recyclerView.adapter = MyAdapter(getData())
        recyclerView.addItemDecoration(object : RecyclerView.ItemDecoration() {
            override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
                super.getItemOffsets(outRect, view, parent, state)
                outRect.set(0, 10, 0, 0)
            }
        })

    }

    fun getData(): List<Bean> {

        var list = mutableListOf<Bean>();
        for (i in 0 until 50) {

            var title = ""
            var content = ""
            var note = ""

            if (i < 20) {
                for (j in 0 until i) {
                    title += "title:$i"
                    content += "content:${i % 6}"
                    note += "note:${i % 4}"
                }
            }


            list.add(Bean(title, content, note))
        }
        return list

    }


    class MyAdapter(var data: List<Bean>) : RecyclerView.Adapter<MyHolder>() {


        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder {
            return MyHolder(LayoutInflater.from(parent.context).inflate(R.layout.recycler_constraint_item, parent, false))
        }

        override fun getItemCount(): Int {
            return data.size
        }

        override fun onBindViewHolder(holder: MyHolder, position: Int) {
            holder.bind(data.get(position), position)
        }

    }

    class MyHolder(view: View) : RecyclerView.ViewHolder(view) {
        var titleView: TextView
        var contentView: TextView
        var noteView: TextView
        var btn: Button

        init {
            titleView = view.findViewById(R.id.tv_title)
            contentView = view.findViewById(R.id.tv_content)
            noteView = view.findViewById(R.id.tv_note)
            btn = view.findViewById(R.id.btn)
        }

        fun bind(bean: Bean, pos: Int) {
            setText(titleView, bean.title)
            setText(contentView, bean.content)
            setText(noteView, bean.note)
            btn.setText("$pos")

            btn.setOnClickListener {
                if (pos % 2 == 0) {
                    bean.title = "title"
                    setText(titleView, bean.title)
                }

                if (pos % 3 == 0) {
                    bean.content = "content"
                    setText(contentView, bean.content)

                }

                if (pos % 4 == 0) {
                    bean.note = "note"
                    setText(noteView, bean.note)
                }
            }

        }

        fun setText(view: TextView, text: String) {
            if (TextUtils.isEmpty(text)) {
                view.visibility = View.GONE
            } else {
                view.visibility = View.VISIBLE
                view.text = text
            }

        }
    }

    data class Bean(var title: String, var content: String, var note: String)
}